### DISCUSSION TITLE
Balancer Settings for 48V LiFePO4 (16x3.2V Cells)

### DISCUSSION THREAD

**Discussion by herrep** [Louisvdw/dbus-serialbattery#189] [Q&A] [answered]
Hi,

Thanks to Louis' great work I got my two battery banks of 16 LiFePO4 3.2 V cells connected to my Victron MultiPlus II 48/5000/70 and controlled by a Cerbo GX. The battery banks are each managed by a JKBMS.

From the [FAQ section](https://github.com/Louisvdw/dbus-serialbattery/wiki/FAQ#why-is-the-max-cell-voltage-set-to-345v) here I learned that the cells should be charged in a range between MIN_CELL_VOLTAGE = 3.10 V and MAX_CELL_VOLTAGE = 3.45 V which seems to correspond to a range of about <10% SOC ... 99% SOC.

This leads me to the following two issues:

1) I wonder how the corresponding BMS settings have to look like so that the BMS processes for balancing the cells match the aforementioned charge voltage range. What is a recommended voltage where the BMS should start balancing their monitored cells? Other users recommended to start balancing at 3.45 V because of the flat curve below this voltage. But if charging stops at 3.45 V how should balancing work? Could anybody post recommendable JKBMS settings (or from another BMS)?

2) I assume that the MIN_CELL_VOLTAGE is measured under discharging conditions while the MAX_CELL_VOLTAGE is measured under charging conditions. MIN_CELL_VOLTAGE = 3.10 V seems to correspond to a very low SOC, below 10%.  What about increasing MIN_CELL_VOLTAGE to 3.20 V to stop discharging at 20% SOC? Or is there a better way of setting the minimum SOC for discharge?

Thank you very much for sharing your experience and recommend suitable settings for the described scenario.

Best regards,
Peter

**Comment by Louisvdw** [comment, marked as answer]
Hi

LiFePO4 cells have a voltage of 3.2V which you can take as their centre point or what they are at at 50%. 

Most BMS will BMS will block charging (over voltage protection) at either 3.65V or some 3.7V ,
And  block discharging at 2.8V (over discharge protection).
Using your battery within those settings so that it does not reach those limits is preferred, and that is why the driver aims for values that are between those. 

It is important to note that the SOC% does not relate to a voltage. Lithium cells are not like lead acid where there is a gradual voltage drop as the battery gets depleted. It has a sudden drop when depleted and a fast spike when full. That is why your BMS will could the energy in and out for the SOC. The cell voltages is just if there are imbalances and you will see the values for this is not evenly spaced. The default for your BMS would work the best in most cases, so I suggest you keep it.

1. If all your cells are at 3.45V the balancer does not need to do anything. The balancer only has to work (as an example) if all cells are at 3.40V and one is at 3.50V. Then it will remove energy from the 3.50V cell until they are all the same again. So we aim for the perfect and the balancer help those cells that are a little bit ahead of the rest. I start my balancers on anything above 3.35V. For most of the charging cycle all the cells will have the same voltage and the balancer will do nothing, even if this is above 3.35V.
What you don't want to happen is that balancing is remove energy from your battery when it is discharging or standing idle, because then the battery will loose storage. For an active balancer like the JKBMS the extra energy is transfered to the lowest cells, while a passive balancer use a resistor to remove the energy and for this they would only balance on charging.

2. SOC is used to stop discharge (or charge) and voltage is not a good indicator of SOC%. We only use the min voltage to stop problems (protection) and that should only triggerif one cell is too low. If all the cells in the battery are the same, then we will reach the min SOC way before such a voltage drop happen.

**Reply by herrep** [reply]
Thank you very much for your detailed answer.

My concern is that starting balancing at a charging voltage well below 3.45 V means that balancing is performed when the cells are far away from "almost full". According to the charging profiles I recorded for my cells, a voltage of 3.35 V corresponds to somewhat about 50-60% SOC.  Clearly, the charging and discharging curve is extremely flat, for example discharging in the voltage range of 3.295 ... 3.305 V currently corresponds to 77 ... 99.2% SOC for my cells.  So 1 mV difference can be more than 20% SOC.  It is very clear that voltage cannot be used to precisely determine SOC.  However, my concern is whether a balancing performed at voltages in the range 3.35 ... 3.45 V really work or whether proper balancing requires to exceed the 3.45 V range.  Of course, I would not do this for each cycle.

As I found quite contradictory approaches for balancing cells, I wonder whether further insights coud be gained by opening a dicussion here.

**Reply by Louisvdw** [reply]
(Note: I considder a LiFePO4 cell at 3.45V as full. There is very little energy stored above that voltage.)

You need to ask yourself: why and how do your BMS do balancing?

The **why** would be so that your cells can reach full capacity at about the same time so that you can get the most useful capacity from those cells. Now if some of those cells have a high voltage and your balancer start to work to reduce that cell voltage to be in line with the rest it is doing it's job.

The **how** is that some energy is removed from the cell that is higher than the others. That is a very small amount that is removed. For most BMS the balancers can remove between 50mA - 600mA (some JKBMS do go up to 1A or 5A so check your model). 

If you only have a small battery and your balancer can react with almost that full charging current, then it does make sense to only activate the balancer once a cell reach overvoltage (3.45V). 
But if you have a large battery and your balancer can only work on a small current compared to your cells you want to give the balancers as much time as possible to react and get your cells the same. For this reason I would say you should start balancing earlier.

Lets look at an example why I say this.
Assume the balancer in my BMS can do 500mA(0.5A) per cell and I have 4x 10Ah cells. If I charge the battery with 10A, each cell will get 2.5A of the charge. If one cell is bein balanced it will get 2.5A going in and 0.5A going out so the cell will still receive a 2A charge. 
If you do the same for a battery from 4x 100Ah cells that can easily be charged at 50A or more that would be 25A per cell which the balancer could reduce to 24.5A and you should start to see why I say you need to give the balancer as much time to work as possible. If your cells are well balanced then there would be very little to do. But as your cells get older they will start to age differently and will need more balancing to be kept in check.

By using a higher balance start voltage it is true that less of the charge energy will be lost to balancing. But when you compare that balancing current to your charge current it is a small fraction of the charge current

**Reply by herrep** [reply]
I have now set the JKBMS (2A balance current) to 3.35 V as voltage for starting balancing my 280 Ah cells. Up to now, the widest voltage spread is 12 mV (highest to lowest voltage), while the majority of the cells in between has only 1 or 2 mV difference. I hope that this settig will keep the cells well balanced.

However, I have another issue resulting from two battery modues connected in parallel, each having 16 cells LiFePO4 3.2 V with 280 Ah.  I have a voltage difference between my first battery module and my second battery module of about 50 mV. Before connecting the two battery modules, I charged all cells (parallel) up to 3.65V and disconnected them at 0.05C=14A. Battery cables are of equal length. Now the first battery module shows 30% SOC while the second battery module is at 17% SOC. This difference increased over a number of cycles.

You drivers seem to apply BMS induced limitations only from the BMS of the first battery module which has the higher SOC. This seems to put some stress on the second battery module, as the first battery module is discharged until it reaches 20% SOC while the second battery suffers from a too low SOC.

Each separate driver seems to identify suitable parameter values when checking the parameters of each of the drivers: Given the above-mentioned SOC values, the first driver asked for a discharge limit of 100 A, while the second driver asked for 50 A due to the lower SOC. However, at the end, the discharge current was limited to 100 A for both modules. We do not need to discuss that two identical battery modules in parallel have to be treated the same way but I would expect that we need to work with the values of the weaker battery. So although the driver of the second battery module defined a limitation of 50 A discharge current, the actual discharge current was close to 100 A.

Is there a way how to improve situation? Could you assume from my problem statement that I have some misconfiguration that I need to sort out?

**Reply by Louisvdw** [reply]
As the driver does not support multi banks yet, only the settings from one bank will be used and there is no way to choose which one. What you can do it to unplug the communication from the battery that you don't want to be used and then the remaining battery will be the master device. 

With regards to the voltage difference that is normally because of the difference in resistance between the two batteries. Electricity, just like water, will go the easier route and if the one battery has more resistance, then less energy will flow there. 
Check the connections between each cell. Also make sure the wires from each battery is the same length and thinkness. 

There is nothing the driver will be able to do to assist with this, even if we support multi banks. If the one battery ask for a 50A charge and the next for 30A then the driver will have to combine that for 80A. However the driver has no way to say that that 80A should be split like that. The electricity will flow to the lowest resistance.

**Reply by herrep** [reply]
By sure the current will flow according to the indirect reciprocity of the battery modules so that the current cannot be limited individually for each battery module, only in total.

Your example of requesting a current limitation to 50 A for the first battery module and 30 A for the second battery module is helpful. However, at present, I still do not exactly know what will happen.

I understand that the optimized solution for an (improved) driver will be to request a total current of 80 A where, of course, the driver cannot influence how the 80 A are split between the two modules. In my case we would end up with about 40 A for each battery module due to the more or less identical setup. But I rather assume that this functionality is not in place.

From what I understood is that one of the drivers "wins" and then this setting is applied. Let us assume it is the first driver with the 50 A limitation. Will be the limitation automatically doubled because of two active battery modules? Or should the two battery modules share the 50 A limitation, resulting in 25 A per battery module?

In my observed case, 200 A were drawn in total, while the first driver requested 100 A and the second driver requested 50 A.  Is it  true that your driver identifies the number of battery modules and then multiplies the current limitation with the number of identified battery modules? This would be very fine.

**Comment by Louisvdw** [comment]
The current driver will use the value from one of the batteries. So in my example above the charge will be either 50A or 30A. The driver does not double anything. It does not know about a second battery and it will think there is only 1.

I future the charge will be 80A where the charge that both batteries are asking for will be added. I might limit this if test show problems, but it will never be more than the sum.

**Comment by herrep** [comment]
Given your example, the current limitation of 50 A or 30 A applied to two batteries would result in an effective limit of 25 A or 15 A, respectively, if two batteries are connected.

I assume that I can simply modify your code to double the current limit values because of two batteries. Like adding a constant NUMBER_OF_BATTERIES = 2 and multiplying this constant with the calculated current limits at an appropriate position in the code. Could you give me a hint where to implement this?

In my previous test, the first battery driver showed a discharge limit of 100 A and the second battery driver showed a discharge limit of 50 A. But in fact the discharge current was about 98 A per battery.  Obviously, the limit was not applied, neither from the first nor from the second driver. According to your last reply, only one setting should have been applied. In my case, the first battery driver is the active one. Because of a limit of 100 A to be applied to both batteries, the current drawn per battery should have been about 50 A. Instead, the 3x MultiPlus II almost went to their maximum of discharging the batteries at about 196 A which is close to the 3x MultiPlus maximum inverter current (210 A = 3 x 70 A).

In the Cerbo GX, I have DVCC enabled and all other settings in the DVCC menu disabled. Furthermore, the MultiPlus II have charge settings of 70 A maximum with remote override ticked. So I wonder what setting I have missed that the discharge current was not limited?

**Reply by Louisvdw** [reply]
The easy way would be to just double your charge and discharge current values in utils.py which would cover your two batteries. So if you have a 30A and 50A then just set the limit as 80A. Which ever BMS is used will set that 80A as the limit


To trace your limits you need to look at the BMS Charge/Discharge Limits graph in VRM for the time in question. That will show the limit the driver set and that was used. (either the 100A or the 50% from what you are saying).
Then you need to match that to what the inverter was doing (AC output power graph). If this is not as you expect then either a setting is wrong. 
Remember also that these discharge limits will only be applied if the system is grid tied and can draw extra power from the grid. If your system is off grid it will supply everything it can when requested.
For the settings you will need to review [item 2. under Settings for your GX](https://github.com/Louisvdw/dbus-serialbattery/wiki/How-to-install#settings-for-your-gx-device). If it looks funny tell us which limits you have set where. (utils.py, DVCC Limit charge, ESS Limit inverter power).

**Comment by herrep** [comment]
Indeed, I have already doubled the values for charge and discharge current to cover the total current of the two batteries in parallel. Therefore, both the charge and discharge current was limited to 200 A so that the maximum is 100 A per battery module.

Due to the different SOC values of the two batteries, the first battery module showed a limit of 1/2 of the maximum = 100 A and the second battery module showed a limit of 1/4 of the maximum = 50 A. Given the fact that the first battery module is the active one, the total current should have been limited to 100 A.  I will check the values in the VRM portal to see if everything works fine now.

**Reply by herrep** [reply]
The first screenshot below shows the situation where the discharge current was limited to 200 A:
<img width="979" alt="Limit to 200 A correct" src="https://user-images.githubusercontent.com/41866019/187667524-350225c5-866d-4455-901b-637d8103249d.png">
It is clear to see that the battery discharge current was limited to a value close to 100 A which is fine due to two battery modules. 2x 100 A = 200 A.

However, a short time later, the discharge current was limit to 100 A:
<img width="985" alt="Limit to 100 A wrong" src="https://user-images.githubusercontent.com/41866019/187667741-090ec3ea-172f-4937-b6e4-9a80b1ae85d3.png">
Unfortunately, the battery discharge current did not drop to 50 A because of two batteries with 2 x 50 A = 100 A. Instead, 100 A were drawn from each of the batteries such that the limited value was ignored.

The DVCC settings are as shown in your screenshot, i.e. DVCC enabled while all other settings in the DVCC menu disabled. Furthermore, the MultiPlus II were configured to allow remote control.

**Reply by Louisvdw** [reply]
That all seem to work like it should. Your DCL is 100A and the actual discharge is 96.69A. I'm not sure why you think that is wrong. Where do you see that each battery received a 100A draw - in the phone app for your BMS?

**Reply by herrep** [reply]
As you know, your battery driver only reports the value of exactly one battery. So when a current of 100 A is shown this value concerns only one battery. The other battery has the same value. I have also touch displays connected to the JK BMS and know that each of the batteries showed 100 A. This is also visible from the widgets, as I have a Lynx Shunt as well that shows the total current over both batteries:
<img width="975" alt="Bildschirmfoto 2022-08-31 um 13 55 59" src="https://user-images.githubusercontent.com/41866019/187672930-14494dae-ea7c-43b7-83d2-8e1253aa5e71.png">

At the beginning, the limit was 200 A which was correctly obeyed as each battery drew 100 A. However, later onwards, the limit was reduced to 100 A.  From what I understand how the limit operation works the MultiPlus II should limit the total current drawn from both battery modules altogether to 100 A. But exactly this did not happen. the current was 100 A per battery module, and as yu can see from the Lynx shunt, the total current remained at 200 A despite the limit to 100 A.

**Reply by Louisvdw** [reply]
OK now that I can see that, it does seem like there is an issue.
But I don't know what could be the problem. The DCL was correclty set, so the driver did it's job. It does seem like that was ignored.
If this is not a off-grid set up then you should not see this. 
Perhaps review all your DVCC settings https://www.victronenergy.com/media/pg/CCGX/en/dvcc---distributed-voltage-and-current-control.html

**Reply by herrep** [reply]
Now DCL works after having uploaded my MultiPlus settings file once again.  After invoking ESS mode "keep battery charged" and waiting a while all cells are perfectly balanced with your suggested settings. Even the two JKBMS now report almost identical SOC for both battery modules in parallel.
